src/writer/agents.py

```diff
@@ -0,0 +1,383 @@
+from abc import ABC, abstractmethod
+import re
+from typing import Any, Dict, List, Optional, Union
+from enum import Enum
+import json
+
+class AgentOutput:
+    """Represents the output from an agent's execution"""
+    def __init__(self, response: str, reasoning: List[str], actions: List[Dict[str, Any]]):
+        self.response = response
+        self.reasoning = reasoning
+        self.actions = actions
+
+class Tool:
```

Comment:
`class Tool` should have a serialisation method.
The standard format looks like this:
```
tools = [
    { 
        "type": "function",
        "function": {
            "name": "calculate_mean", 
            "description": "Calculate the mean (average) of a list of numbers.", 
            "parameters": { 
                "type": "object", 
                "properties": { 
                    "numbers": { 
                        "type": "array", 
                        "items": {"type": "number"}, 
                        "description": "List of numbers"
                    } 
                }, 
                "required": ["numbers"] 
            } 
        }
    }
]
```